Public/Src/Cache/MemoizationStore/Distributed/Stores/RedisMemoizationDatabase.cs

```diff
@@ -0,0 +1,192 @@
+
+namespace BuildXL.Cache.MemoizationStore.Distributed.Stores
+{
+    /// <summary>
+    /// Implements a memoization database through Redis.
+    /// </summary>
+    internal class RedisMemoizationDatabase : MemoizationDatabase
+    {
+        private readonly RedisDatabaseAdapter _redis;
+        private readonly IClock _clock;
+
+        private readonly ObjectPool<StreamBinaryWriter> _writerPool = new ObjectPool<StreamBinaryWriter>(() => new StreamBinaryWriter(), w => w.ResetPosition());
+        private readonly ObjectPool<StreamBinaryReader> _readerPool = new ObjectPool<StreamBinaryReader>(() => new StreamBinaryReader(), r => { });
+
+        /// <inheritdoc />
+        protected override Tracer Tracer => new Tracer(nameof(RedisMemoizationDatabase));
+
+        private string GetKey(Fingerprint weakFingerprint) => $"WF_{weakFingerprint.Serialize()}";
+
+        /// <nodoc />
+        public RedisMemoizationDatabase(
+            RedisDatabaseAdapter redis,
+            IClock clock
+            )
+        {
+            _redis = redis;
+            _clock = clock;
+        }
+
+        /// <inheritdoc />
+        public override Task<Result<bool>> CompareExchange(OperationContext context, StrongFingerprint strongFingerprint, string expectedReplacementToken, ContentHashListWithDeterminism expected, ContentHashListWithDeterminism replacement)
+        {
+            var newReplacementToken = Guid.NewGuid().ToString();
+            return CompareExchange(context, strongFingerprint, expectedReplacementToken, expected, replacement, newReplacementToken);
+        }
+
+        private Task<Result<bool>> CompareExchange(OperationContext context, StrongFingerprint strongFingerprint, string expectedReplacementToken, ContentHashListWithDeterminism expected, ContentHashListWithDeterminism replacement, string newReplacementToken)
+        {
+            return context.PerformOperationAsync(
+                Tracer,
+                async () =>
+                {
+                    var key = GetKey(strongFingerprint.WeakFingerprint);
+                    var replacementMetadata = new MetadataEntry(replacement, _clock.UtcNow.ToFileTimeUtc());
+                    var replacementBytes = SerializeMetadataEntry(replacementMetadata);
+
+                    byte[] selectorBytes = SerializeSelector(strongFingerprint.Selector, isHash: false);
+                    byte[] tokenFieldNameBytes = SerializeSelector(strongFingerprint.Selector, isHash: true);
+
+
+                    var result = await _redis.ExecuteBatchAsync(context, batch => batch.CompareExchangeAsync(key, selectorBytes, tokenFieldNameBytes, expectedReplacementToken, replacementBytes, newReplacementToken), RedisOperation.CompareExchange);
+                    return new Result<bool>(result);
+                });
+        }
+
+        /// <inheritdoc />
+        public override Task<IEnumerable<StructResult<StrongFingerprint>>> EnumerateStrongFingerprintsAsync(OperationContext context)

```

Comment:
Question: Why isn't the operation supported? Is it OK not to support it?

Reply:
The operation is only used for testing, and nothing more. It's ok to not implement this, as long as we disable the corresponding tests that use this. The remaining tests are sufficient to test the functionality of the store.

Reply:
I would probably explain this in a comment, given that the other stores do implement it, but feel free to ignore.

---
In reply to: [320996981](https://github.com/microsoft/BuildXL/pull/832#discussion_r320996981) [](ancestors = 320996981)

Reply:
This is an expensive operation in Redis. Yes, its OK not to support it. There can be unbounded amount of data returned by this call so it is not supported in some cases. Its mostly for some niche uses like externally managed garbage collection.

---
In reply to: [320978949](https://github.com/microsoft/BuildXL/pull/832#discussion_r320978949) [](ancestors = 320978949)